### PR TITLE
[#4625] Prune `MethodCallStatement`s if child `MethodCallExpression` resolves to a compile-time constant.

### LINKED ISSUE #4625: Crash when the result of a `MethodCallStatement` is unused and resolves to a compile-time constant
The following P4 program:
```
header h_t {
    bit<8> f;
}

control C() {
    action bar() {
        h_t h;
        h.minSizeInBits();
    }

    table t {
        actions = { bar; }
        default_action = bar;
    }

    apply {
        t.apply();
    }
}

control proto();
package top(proto p);

top(C()) main;
```
crashes with:
```
Compiler Bug: t.p4(8): visitor returned non-MethodCallExpression type: 8
        h.minSizeInBits();
        ^^^^^^^^^^^^^^^^^
```

### PR DESCRIPTION
**Pull request by kfcripps**
Fixes #4625.

### CODE REVIEW

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 1]
```diff
@@ -3653,6 +3658,7 @@ const IR::Node *TypeInference::postorder(IR::MethodCallExpression *expression) {
                 LOG3("Folding " << mem << " to " << w);
                 if (w < 0) return expression;
                 if (mem->member.name.endsWith("Bytes")) w = ROUNDUP(w, 8);
+                if (getParent<IR::MethodCallStatement>()) return nullptr;
```
Comment (fruffy) [inline frontends/p4/typeChecking/typeChecker.cpp:3661]: This exits early before the type is set, could this have any consequences?

File: frontends/p4/typeChecking/typeChecker.cpp [hunk 2]
```diff
@@ -3615,6 +3615,11 @@ bool TypeInference::onlyBitsOrBitStructs(const IR::Type *type) const {
     return false;
 }
 
+const IR::Node *TypeInference::postorder(IR::MethodCallStatement *mcs) {
+    // Remove mcs if child methodCall resolves to a compile-time constant.
```
[the remark this answers is not in the document]
Comment (kfcripps) [reply frontends/p4/typeChecking/typeChecker.cpp:3619]: AFAIK the only way for `methodCall` to be `nullptr` is if it resolved to a compile-time constant, which means it cannot possibly have side-effects.

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 1 above)
Comment (kfcripps) [reply frontends/p4/typeChecking/typeChecker.cpp:3661]: Why do we need to set a type for a node that is being removed?

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (fruffy) [reply frontends/p4/typeChecking/typeChecker.cpp:3619]: Okay, I would rephrase that the methodcall can only be nullpointer if it is resolved to a compile-time constant.

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 1 above)
Comment (fruffy) [reply frontends/p4/typeChecking/typeChecker.cpp:3661]: I guess this change only applies for the size and valid functions so it should be safe.

**Review by fruffy — APPROVED** [review]

File: frontends/p4/typeChecking/typeChecker.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfcripps) [reply frontends/p4/typeChecking/typeChecker.cpp:3619]: I suppose that one could invent an extern function that returns some compile-time constant AND produces some side-effect(s), but no such function exists. If it did exist, then it would be incorrect for the existing transformations to simply replace its usage with a compile-time constant in the first place, so this PR's changes would not be relevant.

[the remark this answers is not in the document]
Comment (fruffy) [reply frontends/p4/typeChecking/typeChecker.cpp:3619]: Yeah, I was thinking of example like this but then realized that this constant inlining is only possible for a very specific set of functions. General externs with inout arguments or functions are not inlined to a constant iirc. Or not in the type checker at least.

**Comment by kfcripps** [comment]
Thank you @fruffy
